smack-extensions/src/main/java/org/jivesoftware/smackx/usertune/element/UserTuneElement.java

```diff
@@ -0,0 +1,246 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jivesoftware.smackx.usertune.element;
+
+import java.net.URI;
+
+import org.jivesoftware.smack.packet.ExtensionElement;
+import org.jivesoftware.smack.packet.Message;
+import org.jivesoftware.smack.packet.XmlEnvironment;
+import org.jivesoftware.smack.util.NumberUtil;
+import org.jivesoftware.smack.util.XmlStringBuilder;
+
+/**
+ * {@link ExtensionElement} that contains the UserTune.
+ * <br>
+ * Instance of UserTuneElement can be created using {@link Builder#build()} method.
+ * To communicate a meaningful UserTune ,{@link #title} and {@link #source} are set as compulsory parameters.
+ */
+public final class UserTuneElement implements ExtensionElement {
+
+    public static final String NAMESPACE = "http://jabber.org/protocol/tune";
+    public static final String ELEMENT = "tune";
+
+    private final String artist;
+    private final Integer length;
+    private final Integer rating;
+    private final String source;
+    private final String title;
+    private final Integer track;
+    private final URI uri;
+
+    public static final UserTuneElement EMPTY_USER_TUNE = null;
+
+    private UserTuneElement() {
+        this(null);
+    }
+
+    private UserTuneElement(Builder builder) {
+        this.artist = builder.artist;
+        this.length = builder.length;
+        this.rating = builder.rating;
+        this.source = builder.source;
+        this.title = builder.title;
+        this.track = builder.track;
+        this.uri = builder.uri;
+    }
+
+    @Override
+    public String getNamespace() {
+        return NAMESPACE;
+    }
+
+    @Override
+    public String getElementName() {
+        return ELEMENT;
+    }
+
+    @Override
+    public XmlStringBuilder toXML(XmlEnvironment xmlEnvironment) {
+        XmlStringBuilder xml = new XmlStringBuilder(this);
+        if (isEmptyUserTune()) {
+            return xml.closeEmptyElement();
+        }
+        xml.rightAngleBracket();
+        xml.optAppend(getUserTuneXML());
+        return xml.closeElement(getElementName());
+    }
+
+    private CharSequence getUserTuneXML() {
+
+        XmlStringBuilder userTuneXML = new XmlStringBuilder();
+        userTuneXML.optElement("artist", artist);
+        userTuneXML.optElement("length", length);
+        userTuneXML.optElement("rating", rating);
+        userTuneXML.element("source", source);
+        userTuneXML.element("title", title);
+        userTuneXML.optElement("track", track);
+        userTuneXML.optElement("uri", uri);
+        return userTuneXML.toString();
+    }
+
+    private boolean isEmptyUserTune() {
+        if (this.artist == null && this.length == null && this.rating == null && this.source == null && this.title == null
```

Comment:
Ok, you seem to really like this, which is not uncommon for an inexperienced programmer. What you really want to write here is

```
return this.artist == null && this.length == null && this.rating == null && this.source == null && this.title == null 
           && this.track == null && this.uri == null;
```

Or, now that `UserTuneElement` has `hashCode()` and `equals()`, why not simply `this.equals(EMPTY_USER_TUNE)`? A lot more readable, no?